apps/astarte_appengine_api/lib/astarte_appengine_api/device/device.ex

```diff
@@ -991,22 +991,46 @@ defmodule Astarte.AppEngine.API.Device do
       count == 1 ->
         [only_path] = paths
 
-        retrieve_endpoint_values(
-          client,
-          device_id,
-          :object,
-          :datastream,
-          interface_row,
-          endpoint_id,
-          endpoint_row,
-          only_path,
-          opts
-        )
+        {:ok, %Astarte.AppEngine.API.Device.InterfaceValues{data: values, metadata: metadata}} =
+          retrieve_endpoint_values(
+            client,
+            device_id,
+            :object,
+            :datastream,
+            interface_row,
+            endpoint_id,
+            endpoint_row,
+            only_path,
+            opts
+          )
+
+        case values do
+          [] ->
+            %{}
+
+          values_list when is_list(values_list) ->
+            simplified_path = simplify_path(path, only_path)
+
+            case simplified_path do
+              "" ->
+                {:ok, %InterfaceValues{data: values_list, metadata: metadata}}
+
+              _ ->
+                values_map =
+                  %{simplified_path => values_list}
+                  |> MapTree.inflate_tree()
+
+                {:ok, %InterfaceValues{data: values_map, metadata: metadata}}
+            end
+
+          _ ->
+            {:ok, %InterfaceValues{data: values}}
+        end
```

Comment:
When you add the `with` above, try to handle this inside the with, possibly adding small helper functions. Nested `case` are a bit ugly.